File: src/procedures/Procedure.java
package procedures;

import led.ColorRGB;
import led.LEDDataBundle;
import led.LEDStripManager;

public abstract class Procedure {
	public LEDStripManager mStrip;
	public ProcedureCalls mCallbacks;
	
	int mStep = 0;
	int mSteps = 0;
	
	int mModulo = 1;
	boolean mModuloInvert = false;
	
	protected boolean mIsSubProcedure = false;
	
	public Procedure(LEDDataBundle _bundle) {
		mStrip = _bundle.ledStrip;
		mCallbacks = _bundle.procedureCalls;
		
		if (_bundle.puModulo != null) {
			mModulo = _bundle.puModulo;
		}
		
		if (_bundle.puModuloInvert != null) {
			mModuloInvert = _bundle.puModuloInvert;
		}
	}
	
	public abstract void start();
	abstract void update();
	void postUpdate() {
		if(mModulo < 2) return;
		
		for (int i = 0; i < LEDStripManager.LED_COUNT; i++) {
			if (mModuloInvert) {
				if (i % mModulo != 0) {
					mStrip.setPixel(i, ColorRGB.black.toSystemColor());					
				}
			}
			else {
				if (i % mModulo == 0) {
					mStrip.setPixel(i, ColorRGB.black.toSystemColor());					
				}
			}
		}
		
		if (mStep > mSteps) {
			finishProcedure();
		}
	}
	
	protected void finishProcedure(boolean _clearStrip) {
		mStrip.mProcContainer.removeCurrentProcedure();
		if (_clearStrip) {
			mStrip.setAllPixels(ColorRGB.black.toSystemColor());
		}
		mCallbacks.onProcedureFinish();
	}
	
	protected void finishProcedure() {
		mStrip.mProcContainer.removeCurrentProcedure();
		mCallbacks.onProcedureFinish();
	}
}

File: src/procedures/Procedure.kt
package procedures

import led.ColorRGB
import led.LEDDataBundle
import led.LEDStripManager

abstract class Procedure(_bundle: LEDDataBundle) {
    var mStrip: LEDStripManager?
    var mCallbacks: ProcedureCalls?
    var mStep = 0
    open var mSteps = 0
    open var mModulo = 1
    var mModuloInvert = false
    protected var mIsSubProcedure = false
    abstract fun start()
    abstract fun update()
    fun postUpdate() {
        if (mModulo < 2) return
        for (i in 0 until LEDStripManager.LED_COUNT) {
            if (mModuloInvert) {
                if (i % mModulo != 0) {
                    mStrip!!.setPixel(i, ColorRGB.black.toSystemColor())
                }
            } else {
                if (i % mModulo == 0) {
                    mStrip!!.setPixel(i, ColorRGB.black.toSystemColor())
                }
            }
        }
        if (mStep > mSteps) {
            finishProcedure()
        }
    }

    protected fun finishProcedure(_clearStrip: Boolean) {
        mStrip!!.mProcContainer.removeCurrentProcedure()
        if (_clearStrip) {
            mStrip!!.setAllPixels(ColorRGB.black.toSystemColor())
        }
        mCallbacks!!.onProcedureFinish()
    }

    protected fun finishProcedure() {
        mStrip!!.mProcContainer.removeCurrentProcedure()
        mCallbacks!!.onProcedureFinish()
    }

    init {
        mStrip = _bundle.ledStrip
        mCallbacks = _bundle.procedureCalls
        if (_bundle.puModulo != null) {
            mModulo = _bundle.puModulo!!
        }
        if (_bundle.puModuloInvert != null) {
            mModuloInvert = _bundle.puModuloInvert!!
        }
    }
}